### ISSUE TITLE
[Feature Request] Unbiased learning-to-rank tutorial

### ISSUE THREAD

**Issue by Ekkalak-T** [tensorflow/ranking#12]
Could you please provide a tutorial or example how to use this feature when training the model?

**Comment by xuanhuiwang** [comment]
This is a nice feature to have. We will consider this in the future release. So far, suppose you have a per-example weight feature named "weight_feature_string" in your input data (similar to a regular per example feature), you can create the loss as follows:

```
u_loss = make_loss_fn("pairwise_logistic_loss", weights_feature_name="weight_feature_string", reduction=MEAN)
```

Similarly the wights can be enabled in the metrics.

The loss on TensorBoard will be different from the unweighted ones. You may need to tune the learning rate to see how the weight affect your weighted metrics.

**Comment by eggie5** [comment]
I'm using the _SkipAbove_ technique [1] to generate relevance pairs with less bias than the raw click counts. What is the next step to explore in removing bias? 

[1] Joachims. Accurately Interpreting...  2005

**Comment by bendersky** [comment]
Alex, in your system, do you have any ways to randomize or perturb the data that you show the users for some small %-age of traffic? If that's the case, you can use the IPW (inverse propensity weighting) technique we describe in this paper: https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/45286.pdf
It has the advantage of being intuitive and simple to implement.

If you cannot randomize the data you can also try out some EM-based techniques described here: https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/46485.pdf

**Comment by levyfan** [comment]
Hi @xuanhuiwang, there is an interesting paper [1] which models IPW (`weights_feature_name`) as a learnable parameter against document position. Is it practicable to implement this idea in tf-ranking?

[1] Hu, Ziniu & Wang, Yang & Peng, Qu & Li, Hang. (2019). Unbiased LambdaMART: An Unbiased Pairwise Learning-to-Rank Algorithm. 10.1145/3308558.3313447.

**Comment by KimuraTian** [comment]
Hi, @xuanhuiwang,

Tensorflow-ranking looks very good, and we are trying to use it for the unbiased learning-to-rank in research. We have several questions. If you could help us, we would appreciate it.

1. based on the first example code in the 1-click demo, it seems that the easiest way to do propensity-weighted training is to append one more column (propensity-weights) in the input data of LIBSVM format. What the following format should be correct? (Suppose propensity P_1 = 1, P_2 = 0.5, and the third feature column is propensity weight)
a.(only weight clicked items)  
1 qid:1 1:1 2:1 3:1
0 qid:1 1:0 2:0 3:0
0 qid:1 1:0 2:1 3:0
0 qid:2 1:0 2:0 3:0
1 qid:2 1:1 2:0 3:0.5
0 qid:2 1:0 2:0 3:0
-----------------
b.(only weight clicked items but with inversed weights) 
1 qid:1 1:1 2:1 3:1
0 qid:1 1:0 2:0 3:0
0 qid:1 1:0 2:1 3:0
0 qid:2 1:0 2:0 3:0
1 qid:2 1:1 2:0 3:2 # (inversed)
0 qid:2 1:0 2:0 3:0
------------------
c. (same weights for all items in the same qid)  
1 qid:1 1:1 2:1 3:1
0 qid:1 1:0 2:0 3:1
0 qid:1 1:0 2:1 3:1
0 qid:2 1:0 2:0 3:2
1 qid:2 1:1 2:0 3:2 # (inversed)
0 qid:2 1:0 2:0 3:2
------------------

2. in LIBSVM data format, does the scale of target values (e.g. (1, 0) vs 5-star scale) affect the training (if using pairwise_logistic_loss) and the evaluation (e.g. call `ranker.evaluate` to compute ndcg@5)?

3. I saw the evaluation metrics computed by `ranker.evaluate` is not deterministic even for one single run of the same test data if I use `tfr.data.libsvm_generator` as input function. Is there a way to make this deterministic and also to keep items as original positions in the ranking list in input data? Since we want to know the actual items ids in the predicted ranking list after calling `ranker.predict`. If the output were shuffled, it's hard to map them to their ids, especially for LIBSVM input data.

4. Can we output the metrics computed for each test example instead of one single mean score for some statistic test?

Thank you so much!

**Comment by xuanhuiwang** [comment]
@KimuraTian, See blow for the replies to your questions:

1. You should use b.

2. The scale does not affect the pairwise loss since they are based on indicator function y_i > y_j. However, the scale has impact on metrics.

3. Please avoid using libsvm_generator. We should remove that in some version. Please take a look at the ./example directory. I assume that your data can fit in memory if you use LIBSVM format. Otherwise, consider using EIE version in https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/handling_sparse_features.ipynb.

4. This is a good point. Not sure how to do it exactly. You may try to eval over an epoch with N batches and print out the metric for each batch.

**Comment by KimuraTian** [comment]
@xuanhuiwang, Thank you very much! Well answered my questions.

**Comment by xuanhuiwang** [comment]
This commit shows how to incorporate the given weights into unbiased LTR in tf-ranking: https://github.com/tensorflow/ranking/commit/029fffb834fefb6727506ff3db3e9ec262223fb9.

Let us know if you have any questions.

**Comment by eggie5** [comment]
@bendersky just wanted to follow up almost a year later and thank you for the great reference on IPS w/ randomization experiments. Here's a preview of a segmented bias model (across clients):

![segmetned-bias_model-CENSORED](https://user-images.githubusercontent.com/1122/109878197-f4fef480-7c41-11eb-9c59-511c5489499d.png)
